File: app/src/main/java/com/hana053/micropost/interactors/InteractorModule.java
package com.hana053.micropost.interactors;

import com.hana053.micropost.BuildConfig;
import com.hana053.micropost.services.AuthTokenService;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory;
import retrofit2.converter.moshi.MoshiConverterFactory;

@Module
@SuppressWarnings("WeakerAccess")
public class InteractorModule {

    @Provides
    @Singleton
    public OkHttpClient provideOkHttpClient(AuthTokenService authTokenService) {
        return new OkHttpClient().newBuilder()
                .addInterceptor(chain -> {
                    Request original = chain.request();
                    String authToken = authTokenService.getAuthToken();

                    Request.Builder builder = original.newBuilder();
                    if (authToken != null) builder.header("authorization", "Bearer " + authToken);
                    Request request = builder
                            .method(original.method(), original.body())
                            .build();
                    return chain.proceed(request);
                })
                .build();
    }

    @Provides
    @Singleton
    public Retrofit provideRetrofit(OkHttpClient okHttpClient) {
        return new Retrofit.Builder()
                .baseUrl(BuildConfig.API_URL)
                .addConverterFactory(MoshiConverterFactory.create())
                .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                .client(okHttpClient)
                .build();
    }

    @Provides
    @Singleton
    public LoginInteractor provideLoginInteractor(Retrofit retrofit) {
        return retrofit.create(LoginInteractor.class);
    }

}

File: app/src/main/java/com/hana053/micropost/interactors/InteractorModule.kt
package com.hana053.micropost.interactors

import com.github.salomonbrys.kodein.Kodein
import com.github.salomonbrys.kodein.bind
import com.github.salomonbrys.kodein.instance
import com.github.salomonbrys.kodein.singleton
import com.hana053.micropost.BuildConfig
import com.hana053.micropost.services.AuthTokenService
import com.hana053.myapp.interactors.FeedInteractor
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory
import retrofit2.converter.moshi.MoshiConverterFactory

fun interactorModule() = Kodein.Module {

    bind<OkHttpClient>() with singleton {
        val authTokenService = instance<AuthTokenService>()

        OkHttpClient().newBuilder()
            .addInterceptor { chain ->
                val authToken = authTokenService.getAuthToken()
                val original = chain.request()
                val builder = original.newBuilder()
                if (authToken != null)
                    builder.header("authorization", "Bearer " + authToken)
                val request = builder
                    .method(original.method(), original.body())
                    .build()
                chain.proceed(request)
            }
            .build()
    }

    bind<Retrofit>() with singleton {
        val okHttpClient = instance<OkHttpClient>()

        Retrofit.Builder()
            .baseUrl(BuildConfig.API_URL)
            .addConverterFactory(MoshiConverterFactory.create())
            .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
            .client(okHttpClient)
            .build()
    }

    bind<LoginInteractor>() with singleton {
        val retrofit = instance<Retrofit>()
        retrofit.create(LoginInteractor::class.java)
    }

    bind<FeedInteractor>() with singleton {
        val retrofit = instance<Retrofit>()
        retrofit.create(FeedInteractor::class.java)
    }

    bind<UserInteractor>() with singleton {
        val retrofit = instance<Retrofit>()
        retrofit.create(UserInteractor::class.java)
    }

    bind<RelationshipInteractor>() with singleton {
        val retrofit = instance<Retrofit>()
        retrofit.create(RelationshipInteractor::class.java)
    }

    bind<UserMicropostInteractor>() with singleton {
        val retrofit = instance<Retrofit>()
        retrofit.create(UserMicropostInteractor::class.java)
    }

    bind<MicropostInteractor>() with singleton {
        val retrofit = instance<Retrofit>()
        retrofit.create(MicropostInteractor::class.java)
    }

    bind<RelatedUserListInteractor>() with singleton {
        val retrofit = instance<Retrofit>()
        retrofit.create(RelatedUserListInteractor::class.java)
    }

}

